pulp_docker/app/tasks/recursive_remove.py

```diff
@@ -0,0 +1,98 @@
+    unless that content is also related to content that will remain in the
+    repository. For example, if a manifest-list is specified, we need to remove
+    all referenced manifests unless those manifests are referenced by a
+    manifest-list that will stay in the repository.
+
+    Args:
+        repository_pk (int): The primary key for a Repository for which a new Repository Version
+            should be created.
+        content_units (list): List of PKs for :class:`~pulpcore.app.models.Content` that
+            should be removed from the Repository.
+    """
+    repository = Repository.objects.get(pk=repository_pk)
+    latest_version = RepositoryVersion.latest(repository)
+    latest_content = latest_version.content.all() if latest_version else Content.objects.none()
+
+    tags_to_remove = Tag.objects.filter(
+        pk__in=content_units,
+    )
+    tags_to_remain = Tag.objects.filter(
+        pk__in=latest_content.filter(_type='docker.tag')
+    ).exclude(pk__in=tags_to_remove)
+
+    manifest_lists_must_remain = Manifest.objects.filter(
+        pk__in=tags_to_remain.values_list("tagged_manifest", flat=True),
+        media_type=MEDIA_TYPE.MANIFEST_LIST
+    )
+
+    manifest_lists_to_remove = Manifest.objects.filter(
+        pk__in=content_units,
+        media_type=MEDIA_TYPE.MANIFEST_LIST
+    ) | Manifest.objects.filter(
+        pk__in=tags_to_remove.values_list('tagged_manifest', flat=True),
+        media_type=MEDIA_TYPE.MANIFEST_LIST,
+    )
+    manifest_lists_to_remove = manifest_lists_to_remove.exclude(pk__in=manifest_lists_must_remain)
+
+    # This is necessary to determine manifests that are safe to remove.
+    # manifest lists that *must* remain are those which have tags that are not removed.
+    # however, some untagged manifest lists may exist, and if they are not
+    # removed, manifest lists to remain will be bigger than manifest_lists_must_remain
+    manifest_lists_to_remain = Manifest.objects.filter(
+        pk__in=latest_content.filter(_type='docker.manifest'),
+        media_type=MEDIA_TYPE.MANIFEST_LIST,
+    ).exclude(pk__in=manifest_lists_to_remove)
+
+    manifests_must_remain = Manifest.objects.filter(
+        pk__in=tags_to_remain.values_list("tagged_manifest", flat=True),
+        media_type__in=[MEDIA_TYPE.MANIFEST_V1, MEDIA_TYPE.MANIFEST_V1_SIGNED,
+                        MEDIA_TYPE.MANIFEST_V2]
+    ) | Manifest.objects.filter(
+        pk__in=manifest_lists_to_remain.values_list('listed_manifests', flat=True)
+    )
+
+    manifests_to_remove = Manifest.objects.filter(
+        pk__in=content_units,
+        media_type__in=[MEDIA_TYPE.MANIFEST_V1, MEDIA_TYPE.MANIFEST_V1_SIGNED,
+                        MEDIA_TYPE.MANIFEST_V2]
+    ) | Manifest.objects.filter(
+        pk__in=manifest_lists_to_remove.values_list('listed_manifests', flat=True)
+    ) | Manifest.objects.filter(
+        pk__in=tags_to_remove.values_list('tagged_manifest', flat=True),
```

Comment:
I think we need to filter for media_type here as well. Couldn't some manifest lists sneak in here?